server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -46,42 +68,112 @@ public void beforeRefresh() throws IOException {
+
+                                boolean uploadStatus = uploadNewSegments(refreshedLocalFiles);
+                                if (uploadStatus) {
+                                    remoteDirectory.uploadMetadata(
+                                        refreshedLocalFiles,
+                                        storeDirectory,
+                                        indexShard.getOperationPrimaryTerm(),
+                                        segmentInfos.getGeneration()
+                                    );
+                                    localSegmentChecksumMap.keySet()
+                                        .stream()
+                                        .filter(file -> !refreshedLocalFiles.contains(file))
+                                        .collect(Collectors.toSet())
+                                        .forEach(localSegmentChecksumMap::remove);
+                                }
+                            }
+                        } catch (EngineException e) {
+                            logger.warn("Exception while reading SegmentInfosSnapshot", e);
+                        }
+                    } catch (IOException e) {
+                        // We don't want to fail refresh if upload of new segments fails. The missed segments will be re-tried
+                        // in the next refresh. This should not affect durability of the indexed data after remote trans-log integration.
+                        logger.warn("Exception while uploading new segments to the remote segment store", e);
+                    }
                 }
-            });
+            } catch (Throwable t) {
+                logger.error("Exception in RemoteStoreRefreshListener.afterRefresh()", t);
+            }
+        }
+    }
+
+    // Visible for testing
+    boolean uploadNewSegments(Collection<String> localFiles) throws IOException {
+        AtomicBoolean uploadSuccess = new AtomicBoolean(true);
+        localFiles.stream().filter(file -> !EXCLUDE_FILES.contains(file)).filter(file -> {
+            try {
+                return !remoteDirectory.containsFile(file, getChecksumOfLocalFile(file));
+            } catch (IOException e) {
+                logger.info(
+                    "Exception while reading checksum of local segment file: {}, ignoring the exception and re-uploading the file",
+                    file
+                );
+                return true;
+            }
+        }).forEach(file -> {
+            try {
+                remoteDirectory.copyFrom(storeDirectory, file, file, IOContext.DEFAULT);
+            } catch (IOException e) {
+                uploadSuccess.set(false);
+                // ToDO: Handle transient and permanent un-availability of the remote store (GitHub #3397)
+                logger.warn(() -> new ParameterizedMessage("Exception while uploading file {} to the remote segment store", file), e);
+            }
+        });
+        return uploadSuccess.get();
```

Comment:
the upload is happening serially, can we either bulk upload or do concurrent async uploads